app/models/capabilities/scopes/default.rb

```diff
@@ -117,6 +120,32 @@ def default_sql_by_non_member
           WHERE enabled_modules.project_id IS NOT NULL OR "actions".module IS NULL
         SQL
       end
+
+      def default_sql_by_non_member_with_anonymous
+        <<~SQL.squish
+          SELECT DISTINCT
+            actions.id "action",
+            users.id principal_id,
+            projects.id context_id
+          FROM (#{Action.default.to_sql}) actions
+          JOIN "role_permissions" ON "role_permissions"."permission" = "actions"."permission"
+          JOIN "roles" ON "roles".id = "role_permissions".role_id AND roles.builtin = #{Role::BUILTIN_ANONYMOUS}
+          JOIN (#{Principal.visible.not_builtin_without_anonymous.not_locked_without_anonymous.to_sql}) users
+            ON 1 = 1
+          JOIN "projects"
+            ON "projects".active = true
+            AND ("projects".public = true OR EXISTS (SELECT 1
+                                                    FROM members
+                                                    WHERE members.project_id = projects.id
+                                                    AND members.user_id = users.id
+                                                    LIMIT 1))
```

Comment:
The `anonymous role` can only ever be associated to a single user, the singleton instance of `AnonymousUser` which is a special user instance that a request is associated with if no more specific user is fitting. E.g. if I were to browse community.openproject.com without logging in, I'd be doing so as that instance of `AnonymousUser`.

With the way the SQL is currently written, all actual users in addition to the `AnonymousUser` instance will receive the permissions of the `Role::BUILTIN_ANONYMOUS`.

This can be avoided by not joining all the users but only that one instance of `AnonymousUser`. 

The additional check for whether the user is a member in the project (`EXISTS (SELECT 1 ...)`) can be removed from here since an `AnonymousUser` will never be one. The other conditions, project has to be public and active, has to apply, though.